src/esm_runscripts/workflow.py

```diff
@@ -1,49 +1,189 @@
+    order_in_cluster = "sequential"
+    run_only = None
+    skip_chunk_number = None
+    skip_run_number = None
+
+    def __init__(self, phase_name):
+        self.name = phase_name
+
+class UserWorkflowPhase(WorkflowPhase):
+    """A user workflow phase class."""
+    script = None
+    script_dir = None
+    call_function = None
+    env_preparation = None
+
+    def __init__(self, phase_name):
+        self.name = phase_name
+        batch_or_shell = "batch"
+        submit_to_batch_system = False
 
-def skip_cluster(cluster, config):
-    gw_config = config["general"]["workflow"]
-    clusterconf = gw_config["subjob_clusters"][cluster]
-
-    """
-    print(f"run_only {clusterconf.get('run_only', 'Error') }")
-    print(f"skip_chunk_number {clusterconf.get('skip_chunk_number', -999)}")
-    print(f"skip_run_number {clusterconf.get('skip_run_number', -999)}")
-    print(f"chunk_number {config['general'].get('chunk_number', -998)}")
-    print(f"run_number {config['general'].get('run_number', -998)}")
-    print(f"last_run_in_chunk {config['general']['last_run_in_chunk']}")
-    print(f"first_run_in_chunk {config['general']['first_run_in_chunk']}")
+def assemble_workflow(config):
+    from . import Workflow
     """
+    Assembles the workflow tasks.
+    Is called from the plugin recipe prepcompute.
 
-    if clusterconf.get("run_only", "Error") == "last_run_in_chunk" and not config[
-        "general"
-    ].get("last_run_in_chunk", False):
-        return True
-    if clusterconf.get("run_only", "Error") == "first_run_in_chunk" and not config[
-        "general"
-    ].get("first_run_in_chunk", False):
-        return True
-    if clusterconf.get("skip_chunk_number", -999) == config["general"].get(
-        "chunk_number", -998
-    ):
-        return True
-    if clusterconf.get("skip_run_number", -999) == config["general"].get(
-        "run_number", -998
-    ):
-        return True
-
-    return False
-
-
-def assemble_workflow(config):
-    #
-    config = init_total_workflow(config)
-    config = collect_all_workflow_information(config)
-    config = complete_clusters(config)
-    config = order_clusters(config)
-    config = prepend_newrun_job(config)
+    Arguments:
+        config -- dictionary
+    Returns:
+        config
+    """
 
+    # 1. Generate default workflow object
+    #TODO: preset of default workflow phases should be set in some config file.
+    workflow = Workflow(["prepcompute","compute","tidy"],always_run_with=["prepare","prepexp"])
+    # initialize the default workflow as Workflow object
+    # TODO: NW where are these default phases defined???
+    # 2. Initialize default workflow phases
+    workflow = init_default_workflow(workflow, config)
+    # 3. Read in workflows from runscript and config files
+    workflow = collect_all_user_workflows(workflow, config)
+
+    #config = collect_all_workflow_information(config)
+
+# Why do I need to do the following function call?
```

Comment:
I could only review until here today. I'll try a bit more tomorrow.

Reply:
Great. Thank you very much for your review so far.